### PR TITLE
feature/aggregate-metrics

### PR DESCRIPTION
**Pull request by msadovyi**

### CODE REVIEW

**Comment by valcker** [comment]
Fix eslint errors.

**Comment by valcker** [comment]
Do you use prettier in your editor?

File: packages/backend/src/data/schema/network-metric.schema.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import * as mongoose from 'mongoose';
+import INetworkMetric from '../../common/interface/network-metric.interface';
+
+const NetworkMetricSchema = new mongoose.Schema(
```
Comment (valcker) [inline packages/backend/src/data/schema/network-metric.schema.ts:4]: eslint errors

File: packages/backend/src/data/schema/request-metric.schema.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import * as mongoose from 'mongoose';
+import IRequestMetric from '../../common/interface/request-metric.interface';
+
+const RequestMetricSchema = new mongoose.Schema(
```
Comment (valcker) [inline packages/backend/src/data/schema/request-metric.schema.ts:4]: eslint errors

File: packages/backend/src/test/unit/metric.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+      ]
+      const resolvers = [{ path: 'path', latency: 100500 }];
+
+      const spyCreate = jest.spyOn((metricService as any).requestMetricModel, 'create').mockImplementation();
+      const spyLrange = jest.spyOn((metricService as any).redis, 'lrange').mockResolvedValue(records.map(r => JSON.stringify(r)));
+      const spyLtrim = jest.spyOn((metricService as any).redis, 'ltrim').mockImplementation(() => {});
+      const spyReduceResolvers = jest.spyOn((metricService as any), 'reduceResolvers').mockReturnValue(resolvers);
+
+      const requestId = 1;
+      await (metricService as any).aggregateRequestMetric(requestId);
```
Comment (valcker) [inline packages/backend/src/test/unit/metric.service.spec.ts:130]: type mismatch warning

File: packages/backend/src/test/unit/metric.service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+          path,
+          latency: resolverDoneDate - resolverCalledDate,
+          [MetricsChannels.RESOLVER_CALLED]: resolvers[0],
+          [MetricsChannels.RESOLVER_DONE]: resolvers[1],
+          [MetricsChannels.RESOLVER_ERROR]: resolvers[2],
+        }
+      ])
+    });
+  });
+});
```
Comment (valcker) [inline packages/backend/src/test/unit/metric.service.spec.ts:198]: fix eslint and prettier problems

File: packages/backend/src/test/unit/metric.service.spec.ts (hunk 3 above)
Comment (msadovyi) [reply packages/backend/src/test/unit/metric.service.spec.ts:130]: aggregateRequestMetric it's private method, is there another way for testing besides "any"?

Comment (valcker) [reply packages/backend/src/test/unit/metric.service.spec.ts:130]: it's actually about the argument of the function which accepts string, but `requestId` is number.

File: packages/backend/src/common/interface/network-metric.interface.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+export default interface INetworkMetric {
+  nodeId: string,
+  bytesIn: number,
+  bytesOut: number,
+  connections: number,
+  recordedAt: number,
+}
```
Comment (germankrause) [inline packages/backend/src/common/interface/network-metric.interface.ts:7]: no new line, `;` instead `,`, use prettier

File: packages/backend/src/modules/metric/metric.service.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+  ) {
+    [this.redis] = this.redisClients;
+  }
+
+  private onModuleInit(): void {
+    this.init();
+  }
+
+  public init(): void {
+    const { enabled = true, chunk = 10, delay = 5000 } = config.application.metrics;
```
Comment (germankrause) [inline packages/backend/src/modules/metric/metric.service.ts:39]: why chunk = 100 in default config and 10 here?

File: packages/backend/src/modules/metric/metric.service.ts [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    @Inject(Provider.REDIS) private readonly redisClients: [Redis, Redis],
+    @InjectModel('RequestMetric') private requestMetricModel: Model<IRequestMetricDocument>,
+    @InjectModel('NetworkMetric') private networkMetricModel: Model<INetworkMetricDocument>,
+    private readonly logger: LoggerService,
+  ) {
+    [this.redis] = this.redisClients;
+  }
+
+  private onModuleInit(): void {
+    this.init();
```
Comment (germankrause) [inline packages/backend/src/modules/metric/metric.service.ts:35]: why init body is not here?

File: packages/backend/src/modules/metric/metric.service.ts [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+  AnyMetric,
+  AnyResolverMetric,
+  IGotError,
+  IGotRequest,
+  ISentResponse,
+  IReducedResolver,
+} from './interfaces';
+
+@Injectable()
+export default class MetricService {
```
Comment (germankrause) [inline packages/backend/src/modules/metric/metric.service.ts:21]: run prettier

File: packages/backend/src/test/unit/metric.service.spec.ts [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+  describe('init', () => {
+    it('should call fetchMetrics after delay time', async () => {
+      const spyFetchMetrics = jest.spyOn(metricService as any, 'fetchMetrics').mockImplementation(() => { });
+
+      metricService.init();
+      jest.advanceTimersByTime(config.application.metrics.delay);
+
+      expect(spyFetchMetrics).toBeCalledTimes(2);
+      expect(spyFetchMetrics).nthCalledWith(1, MetricsChannels.REQUEST_IDS, config.application.metrics.chunk);
```
Comment (germankrause) [inline packages/backend/src/test/unit/metric.service.spec.ts:44]: nthCalledWith 👍

File: packages/backend/src/test/unit/metric.service.spec.ts [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+        const spyGetRecords = jest.spyOn((metricService as any), 'getRecords').mockResolvedValue(records);
+        const spyAggregateRequestMetric = jest.spyOn((metricService as any), 'aggregateRequestMetric').mockImplementation();
+
+        await (metricService as any).fetchMetrics(MetricsChannels.REQUEST_IDS, 10);
+
+        expect(spyGetRecords).toBeCalledTimes(1);
+        expect(spyGetRecords).toBeCalledWith(MetricsChannels.REQUEST_IDS, chunk);
+        expect(spyAggregateRequestMetric).toBeCalledTimes(records.length);
+        expect(spyAggregateRequestMetric.mock.calls[0][0]).toBe(records[0]);
+        expect(spyAggregateRequestMetric.mock.calls[1][0]).toBe(records[1]);
```
Comment (germankrause) [inline packages/backend/src/test/unit/metric.service.spec.ts:63]: nthCalledWith for all mock.calls[n]?

File: packages/backend/package.json [hunk 11]
```diff
@@ -13,12 +13,12 @@
     "start:debug": "nest start --debug --watch",
     "start:prod": "node dist/main",
     "lint": "eslint '{src,apps,libs,test}/**/*.ts' --fix",
-    "test": "jest --coverage",
+    "test": "jest --coverage --detectOpenHandles",
```
Comment (germankrause) [inline packages/backend/package.json:16]: let's use detectOpenHandles only while local development, as it slows down the CI

**Review by germankrause — CHANGES_REQUESTED** [review]
resolve conflicts and comments

File: packages/backend/config/env/production.json [hunk 12]
```diff
@@ -8,7 +8,12 @@
       "debug": false
     },
     "logLevel": "@@LOG_LEVEL",
-    "secret": "@@SECRET"
+    "secret": "@@SECRET",
```
Comment (germankrause) [inline packages/backend/config/env/production.json:11]: we don't use secret, let's remove it everywhere

File: packages/backend/src/common/interface/request-metric.interface.ts [hunk 13]
```diff
@@ -6,13 +6,21 @@ export default interface IRequestMetric {
...
   userAgent: string;
-  ip: string
-  resolvers: IReducedResolver[],
-  request: Request,
-  rawResponseBody: string,
-  contentLength: number,
-  error: Error | string | null,
+  ip?: string;
+  resolvers: IReducedResolver[];
+  geo: {
```
Comment (germankrause) [inline packages/backend/src/common/interface/request-metric.interface.ts:11]: `geo?:` ?

File: packages/backend/src/modules/metric/metric.service.ts [hunk 14]
```diff
@@ -31,8 +33,9 @@ export default class MetricService {
     [this.redis] = this.redisClients;
   }
 
-  private onModuleInit(): void {
+  private async onModuleInit(): Promise<void> {
     this.init();
+    await this.setMaxmindClient();
```
Comment (germankrause) [inline packages/backend/src/modules/metric/metric.service.ts:38]: let's run this first, before init

File: packages/backend/src/modules/metric/metric.service.ts (hunk 7 above)
Comment (msadovyi) [reply packages/backend/src/modules/metric/metric.service.ts:37]: it's more clear to have dedicated func, onModuleInit can have many calls inside

File: packages/backend/src/modules/metric/metric.service.ts (hunk 8 above)
Comment (msadovyi) [reply packages/backend/src/modules/metric/metric.service.ts:22]: nothing has changed here

File: packages/backend/src/common/interface/request-metric.interface.ts [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+  query: string;
+  userAgent: string;
+  ip?: string;
+  resolvers: IReducedResolver[];
+  geo: {
+    city?: string;
+    country?: string;
+    location?: {
+      latitude: string;
+      longitude: string;
```
Comment (valcker) [inline packages/backend/src/common/interface/request-metric.interface.ts:16]: can we also extract timezone?

**Review by germankrause — APPROVED** [review]

**Review by valcker — APPROVED** [review]

File: packages/backend/src/common/interface/request-metric.interface.ts (hunk 15 above)
Comment (msadovyi) [reply packages/backend/src/common/interface/request-metric.interface.ts:16]: There is type mismatch, thanks.

**Review by haletsky — APPROVED** [review]

**Pull request merged**